### ISSUE TITLE
thingShadow.subscribe & event (message)

### ISSUE THREAD

**Issue by linqOxelar** [aws/aws-iot-device-sdk-js#44]
Hi

I tried to use thingShadow.subscribe and nothing happened ! no error but no handling of message from topic ! The thingShadow can update, register, unregister, get the Shadow... it can also publish on topics that are not $aws/things... !
But no event.on('message'..), no on.('error')...

Could come from code ? or do i miss something about AWS IOT / mqtt?

Dham

**Comment by gkwicker** [comment]
Hi,

Are you able to run [this example](https://github.com/aws/aws-iot-device-sdk-js/blob/master/examples/thing-passthrough-example.js)?  It uses the thing shadow class publish() and subscribe() methods to do 'plain' MQTT while also performing thing shadow operations.  Instructions for running the example programs can be found [here](https://github.com/aws/aws-iot-device-sdk-js#programs).  Like most of the example programs, you run two copies which talk to one another, either on the same or different hosts.  If this example program works for you, you can use it as a starting point when writing your own application; if not, please let us know.  Thanks for using AWS IoT.

**Comment by dhamfr** [comment]
Hello

After some tuning to make it work (disabling the cmdLineProcess that had some troubles to find my certificates), the result is : 
1. registering a thing already registered seems to disconnect it
2. i tune the reconnectPeriod (10 000) and operationTimeout (20 000) to be able to get 'updated state to thing shadow' on testmode2
3. messages from testmode1 : reconnect / connected / thing already registered / offline / close / reconnect...
4. message from testmode2: reconnect/connected + registering/thing already registered/update still in progress/timout/got 'accepted' status / updated state / offline / close / reconnect
5. so i put reconnectPeriod (30 000) and operationTimeout (50 000) 

Following topics on mqtt.fx i saw no delta, so i delete the Shadow and create one with a reported value different from the current one and then i received message on shadow/delta and only one message on nonThingTopic (on testmode1), but no other message!

Then i tried with operationTimeout (50 000) , but still no nonThingTopic!

Actually i got nonThingTopic message only when i update the shadow from the console !

But neither testmode1 or testmode2 received a message (event:message, handleMessage) ! I saw the nonThingTopic message only on mqtt.fx

Conclusion : 
1. operationTimout seems to be tuned (maybe my network is slow... but never got this trouble before with my own code and i never code any TimeOut)
2. things do not get any messages from subscription on nonThingTopic !!

**Comment by gkwicker** [comment]
Hi,

Can you check to make sure that both of your processes are using different client IDs?  AWS IoT requires that each client use a unique client ID value;  the 'offline/close/reconnect' messages you are seeing on both processes indicate that they are disconnecting and reconnecting to the service, and this is often caused by duplicate client IDs.  The example programs use the '-i' flag to set the client ID, and you can use '--debug=true' to print out debugging information which may help isolate the cause.

**Comment by dhamfr** [comment]
Hello

Ok working with different clientID names and i tried with differents certificates/things : that is working !!

I note that you are setting operationTimeOut in thingShadow#register, which is not documented. You documented it in thingShadow

So i tested different step changes from your code up to mine : things stopped working when i added a callback for thingShadow.subscribe !!
So i dug a bit and that was working if options are added before the callback.

To sum up :

**WORKING**

```
thingShadow.on('connect', function(){
     thingShadow.register(thingName);
     thingShadow.subscribe('nonThingTopic');
}); 
```

**WORKING**

```
thingShadow.on('connect', function(){
     thingShadow.register(thingName);
     thingShadow.subscribe('nonThingTopic', {qos:0}, function(err, granted){
          //qos:0 or 1; 2 not accepted byt AWS IOT i think
          // some code
     });
}); 
```

**NOT WORKING**

```
thingShadow.on('connect', function(){
     thingShadow.register(thingName);
     thingShadow.subscribe('nonThingTopic', function(err, granted){
          // some code
     });
}); 
```

Thanks a lot for help and guidance!

**Comment by chaurah** [comment]
Hi Dham,
      The  subscribe API that we provide uses the same format as provided by the underlying MQTT client. That is why only the two “working” samples you have provided are correct. You can find the reference for that client here - https://github.com/mqttjs/MQTT.js#subscribe
       
      Great to hear that the code is working for you now with different ClientIDs. Please let us know if you have any further questions. Thanks for using AWS IoT.

**Comment by dhamfr** [comment]
Hi Chaurah

Thank you for all information and AWS support.

So now i am discovering that my tasks will not be much about understanding technical specifications and verifying whether my cases are working; they will likely be more of discovering errors in documentation.

if mqtt.Client#subscribe(topic/topic array/topic object, [options], [callback]) and awsIot.thingShadow#subscribe(topic, [options], [callback]) are not correct, according to what i tested they should be :

```
awsIot.thingShadow#subscribe(topic [, options [, callback] ])
mqtt.Client#subscribe(topic/topic array/topic object, [, options [, callback] ])
```

Thank you for helping me to smoothly use AWS IOT

**Comment by gkwicker** [comment]
Hi Dham,

You're correct – the documentation says that these two methods are identical, but actually there is a slight difference.  We will update the documentation in an upcoming release to make this more clear.  Thanks again for using AWS IoT.

**Comment by dhamfr** [comment]
Well, sorry to say that : i have not seen any changes!

**Comment by liuszeng** [comment]
Addressed in release [v1.0.13](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.13).
